### ISSUE TITLE
Think about chances for optimizations when clicking a link during a visit

### ISSUE THREAD

**Issue by hirasso** [swup/swup#811]
**Describe the problem**

[This discussion](https://github.com/orgs/swup/discussions/809) brought it up one more time: Clicking another link during a running page visit [now works](https://github.com/swup/swup/pull/464), but breaks currently running animations. I'd like to get together in a call sometime to think about possible better solutions than what we have now.

**Describe the proposed solution**

<del>

Very rough idea, just as a starting point:

1. Create some kind of "parent" for each page visit, for example a class `VisitController`
2. That controller could have a method `abort` or `disconnect` that would be called each time a new visit was triggered
3. When disconnected, the controller:
    - wouldn't trigger hooks
    - wouldn't alter the dom
    - would completely destroy itself as soon as it can

</del>

The above doesn't make sense at all. I don't know what I was thinking 🤡 

**Notes**

- would this even work with the css-based animation logic of swup?
- if we find out that it could actually work better, could we utilize the existing `Visit` object for it?

**How important is this feature to you?**

- [x] Nice to have. I'd just like to rethink it and document our thoughts about it.

**Comment by daun** [comment]
Sounds like a great topic to explore on a call sometime 🔮

**Comment by daun** [comment]
I guess we could always expand [the current checks](https://github.com/swup/swup/blob/master/src/modules/navigate.ts#L149C1-L152C5) we're doing to identify and ignore visits that have been superseded by a new one in the meantime.

```ts
// Abort if another visit was started in the meantime
if (visit.id !== this.visit.id) {
  return false;
}
```

**Comment by hirasso** [comment]
That will be one part of the solution for sure – I'm unsure about these two:

1. how do we cleanly cancel/restart CSS animations? Something like 
    1. remove animation classes
    2. wait one tick
    3. initialize the next visit
2. how would we go about it if the user clicks another link during the **out**-animation?

Does animating still make sense in this scenario at all, if we think about special animations for routes (Route Name Plugin, JS Plugin)? My initial thought was that the cleanest approach would be to cancel eventually running animations alltogether.

**Comment by daun** [comment]
Regarding 1, that sounds perfectly reasonable. 

On point 2, that's already solved, at least when using CSS class-based transitions. The above check `visit.id !== this.visit.id` will effectively just drop any visits if during the out-animation another visit was requested. Adding the css class for the out-animation again will not really do anything. For clicks during the in-animation, this will not yet work, as well as for JS animations, as these do get repeated, I think.

I agree that dropping any existing visits and short-circuiting would be best, i.e. cancel running animations. This would probably only work if CSS transitions are bound to `html.is-changed`, otherwise we don't have a way of safely removing the transition from the element.

**Comment by hirasso** [comment]
Part 2 is not completely solved I think. The animation promise of a new visit during the out animation will retrieve the wrong animation duration, since it will just read the transition/animation duration from the element, ignoring already elapsed animation time. We'd need to somehow detect the current process of the animation somehow and subtract that from the duration, don't we?

**Comment by hirasso** [comment]
Or are we listening for the `transitionend`/`animationend` event? I'll need to revisit the source code 😅

**Comment by daun** [comment]
We‘re listening for `transitionend` so I think we‘re covered. There’s a timeout there too, but that‘s a fallback if there‘s more or less properties transitioning than expected.

**Comment by hirasso** [comment]
While investigating I found that `is-leaving` is not part of `swupClasses` and thus won't be cleared on `classes.clear()`. Just leaving this here to make sure this is the intended behavior.

**Comment by hirasso** [comment]
Another find: This line saves the visit *by reference* right now, making the comparison for the id later in the function useless:

https://github.com/swup/swup/blob/a8ab81e686bcfb930e634e370e82767e445a58b7/src/modules/navigate.ts#L68

**Comment by daun** [comment]
Solved in #836 and landing soon ☺️
